### PR TITLE
span: fix corruption bug in btree span frontier

### LINKED ISSUE #120987: util/span: btree based Frontier can drop `Forward` calls
PCR has been observed to get stuck in some cases, as described in #120231. While some performance-related issues were identified in that ticket, another issue was identified where the frontier would often become incorrect, according to the updates that were written to it.

One clue however was that this behavior was only observed when using the btree-based frontier; the LLBR frontier would reliably stay synchronized with its rangefeed. Playing back a recording of the rangefeed events in a unit test to an LLBR and BTree frontier side-by-side showed the same behavior, with the btree frontier becoming de-synchronized with the LLBR frontier after between 2k-80k updates.

Jira issue: CRDB-36990

### PR DESCRIPTION
**Pull request by jeffswenson**
Previously, the btree span frontier implementation modified the EndKey of btree entries. The btree internally expects the EndKey to be immutable, so it caused queries for containing spans to return incorrect results.

Now, entries are removed from the btree before their end key is modified. This is expected to have minimal impact on performance because the fast/common path does not require merging or splitting entries.

Fixes: #120987

### CODE REVIEW

**Comment by jeffswenson** [comment]
```
➜  go test . --bench=BenchmarkFrontier
goos: darwin
goarch: arm64
pkg: github.com/cockroachdb/cockroach/pkg/util/span
BenchmarkFrontier/btree=false/rnd-10              834922              1341 ns/op             691 B/op         16 allocs/op
BenchmarkFrontier/btree=false/r=128-10             44199             26000 ns/op            5565 B/op        101 allocs/op
BenchmarkFrontier/btree=false/r=1024-10             4670            249546 ns/op           42526 B/op        721 allocs/op
BenchmarkFrontier/btree=false/r=4096-10             1012           1150424 ns/op          174038 B/op       2852 allocs/op
BenchmarkFrontier/btree=false/r=8192-10              495           2371999 ns/op          349876 B/op       5615 allocs/op
BenchmarkFrontier/btree=false/r=16384-10             236           5053206 ns/op          715567 B/op      11252 allocs/op
BenchmarkFrontier/btree=true/rnd-10              2163332               554.5 ns/op             0 B/op          0 allocs/op
BenchmarkFrontier/btree=true/r=128-10              99777             11327 ns/op               0 B/op          0 allocs/op
BenchmarkFrontier/btree=true/r=1024-10              9013            127850 ns/op               0 B/op          0 allocs/op
BenchmarkFrontier/btree=true/r=4096-10              1990            569113 ns/op               0 B/op          0 allocs/op
BenchmarkFrontier/btree=true/r=8192-10               826           1410739 ns/op               0 B/op          0 allocs/op
BenchmarkFrontier/btree=true/r=16384-10              433           2643127 ns/op               0 B/op          0 allocs/op
PASS
ok      github.com/cockroachdb/cockroach/pkg/util/span  41.653s

➜  git checkout jeffswenson-fix-btree-span-frontier
➜  go test . --bench=BenchmarkFrontier
goos: darwin
goarch: arm64
pkg: github.com/cockroachdb/cockroach/pkg/util/span
BenchmarkFrontier/btree=false/rnd-10              842134              1349 ns/op             693 B/op         16 allocs/op
BenchmarkFrontier/btree=false/r=128-10             47208             26647 ns/op            5688 B/op        103 allocs/op
BenchmarkFrontier/btree=false/r=1024-10             4950            250329 ns/op           42245 B/op        715 allocs/op
BenchmarkFrontier/btree=false/r=4096-10             1072           1146503 ns/op          169264 B/op       2786 allocs/op
BenchmarkFrontier/btree=false/r=8192-10              523           2319801 ns/op          338960 B/op       5466 allocs/op
BenchmarkFrontier/btree=false/r=16384-10             242           4700627 ns/op          653660 B/op      10352 allocs/op
BenchmarkFrontier/btree=true/rnd-10              1850320               648.5 ns/op             0 B/op          0 allocs/op
BenchmarkFrontier/btree=true/r=128-10             103184             11768 ns/op               0 B/op          0 allocs/op
BenchmarkFrontier/btree=true/r=1024-10              9547            127218 ns/op               0 B/op          0 allocs/op
BenchmarkFrontier/btree=true/r=4096-10              2056            591642 ns/op               0 B/op          0 allocs/op
BenchmarkFrontier/btree=true/r=8192-10               858           1403478 ns/op               0 B/op          0 allocs/op
BenchmarkFrontier/btree=true/r=16384-10              445           2640383 ns/op               0 B/op          0 allocs/op
```

**Comment by jeffswenson** [comment]
```
[nix-shell:~/Code/cockroach]$ benchstat before.txt after.txt
goos: darwin
goarch: arm64
                            │ before.txt  │              after.txt              │
                            │   sec/op    │   sec/op     vs base                │
Frontier/btree=true/rnd       647.9n ± 3%   757.4n ± 1%  +16.91% (p=0.000 n=10)
Frontier/btree=true/r=128     13.42µ ± 3%   13.53µ ± 5%        ~ (p=0.912 n=10)
Frontier/btree=true/r=1024    154.3µ ± 1%   151.6µ ± 2%   -1.74% (p=0.004 n=10)
Frontier/btree=true/r=4096    704.1µ ± 1%   704.8µ ± 4%        ~ (p=0.579 n=10)
Frontier/btree=true/r=8192    1.742m ± 1%   1.684m ± 3%   -3.34% (p=0.001 n=10)
Frontier/btree=true/r=16384   3.305m ± 1%   3.263m ± 4%        ~ (p=0.165 n=10)
geomean                       132.6µ        134.9µ        +1.69%

                            │  before.txt  │              after.txt              │
                            │     B/op     │    B/op     vs base                 │
Frontier/btree=true/rnd       0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
Frontier/btree=true/r=128     0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
Frontier/btree=true/r=1024    0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
Frontier/btree=true/r=4096    0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
Frontier/btree=true/r=8192    0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
Frontier/btree=true/r=16384   0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
geomean                                  ²               +0.00%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean

                            │  before.txt  │              after.txt              │
                            │  allocs/op   │ allocs/op   vs base                 │
Frontier/btree=true/rnd       0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
Frontier/btree=true/r=128     0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
Frontier/btree=true/r=1024    0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
Frontier/btree=true/r=4096    0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
Frontier/btree=true/r=8192    0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
Frontier/btree=true/r=16384   0.000 ± 0%     0.000 ± 0%       ~ (p=1.000 n=10) ¹
geomean                                  ²               +0.00%                ²
¹ all samples are equal
² summaries must be >0 to compute geomean
```

**Review by dt — APPROVED** [review]
Nice!

File: pkg/util/span/frontier.go [hunk 1]
```diff
@@ -109,7 +109,7 @@ func (r OpResult) asBool() bool {
 type Operation func(roachpb.Span, hlc.Timestamp) (done OpResult)
 
 var useBtreeFrontier = envutil.EnvOrDefaultBool("COCKROACH_BTREE_SPAN_FRONTIER_ENABLED",
-	metamorphic.ConstantWithTestBool("COCKROACH_BTREE_SPAN_FRONTIER_ENABLED", false))
+	metamorphic.ConstantWithTestBool("COCKROACH_BTREE_SPAN_FRONTIER_ENABLED", true))
```
Comment (msbutler) [inline pkg/util/span/frontier.go:112]: hold up: changing this implies that on 24.3, we'll use the btree frontier right? Should we keep it on for roachtests for a few months before releasing this?

Comment (dt) [reply pkg/util/span/frontier.go:112]: We had it behind a metamorphic for a year now, and had it on by default for a few months already, before I flipped it off in the spring when I tracked down this most recent bug, so anything the roachtests were gonna find they already found. I don't think we learn anything from a delay.

Comment (msbutler) [reply pkg/util/span/frontier.go:112]: good point. Seems fine to merge it now. Hopefully all these scale tests + drt bake time will catch anything else.

**Comment by jeffswenson** [comment]
Thanks for the reviews!

bors r+

**Pull request merged**